src/test/crimson/test_messenger_thrash.cc

```diff
@@ -0,0 +1,668 @@
+      }).then([&test_msg] {
+      return seastar::do_until([&test_msg] {
+        return test_msg.clients_count() == test_msg.clients; },
+      [&test_msg] {
+        return test_msg.init_client(entity_name_t::CLIENT(-1),
+                             "client", get_nonce());
+	});
+      }).then([] {
+        return seastar::when_all_succeed(
+          local_conf().set_val("ms_inject_socket_failures", "30"),
+          local_conf().set_val("ms_inject_internal_delays", "0.1"),
+          local_conf().set_val("ms_inject_delay_probability", "1"),
+	  local_conf().set_val("ms_inject_delay_max", "5"));
+      }).then_unpack([] {
+	return seastar::now();
+      }).then([&test_msg, tp] {
+        return test_msg.generate_connections(tp.connections);
+      }).then([&test_msg, tp] {
+        return test_msg.random_op(tp.random_op);
+      }).then([&test_msg] {
+	  return test_msg.wait_for_done();
+      }).then([] {
+        logger().info("test_inejction() DONE");
+	return seastar::now();
+      }).then([] {
+        return reset_conf();
+      }).handle_exception([] (auto eptr) {
+        logger().error(
+          "test_injection() failed: got exception {}",
+          eptr);
+        throw;
+      });
+    });
+}
+
+}
+
+seastar::future<int> do_test(seastar::app_template& app)
+{
+  std::vector<const char*> args;
+  std::string cluster;
+  std::string conf_file_list;
+  auto init_params = ceph_argparse_early_args(args,
+                                              CEPH_ENTITY_TYPE_CLIENT,
+                                              &cluster,
+                                              &conf_file_list);
+  return crimson::common::sharded_conf().start(init_params.name, cluster)
+  .then([conf_file_list] {
+    return local_conf().parse_config_files(conf_file_list);
+  }).then([&app] {
+    auto&& config = app.configuration();
+    verbose = config["verbose"].as<bool>();
+      return test_stress(thrash_params_t{8, 32, 50, 120})
+    .then([] {
+      return test_injection(thrash_params_t{16, 32, 50, 120});
+    }).then([] {
+      logger().info("All tests succeeded");
+      // Seastar has bugs to have events undispatched during shutdown,
+      // which will result in memory leak and thus fail LeakSanitizer.
+      return seastar::sleep(100ms);
+    });
+  }).then([] {
+    return crimson::common::sharded_conf().stop();
+  }).then([] {
+    return 0;
+  }).handle_exception([] (auto eptr) {
+    logger().error("Test failed: got exception {}", eptr);
+    return 1;
+  });
+}
+
+int main(int argc, char** argv)
+{
+  seastar::app_template app;
+  app.add_options()
+    ("verbose,v", bpo::value<bool>()->default_value(false),
+     "chatty if true");
+  return app.run(argc, argv, [&app] {
+    // This test normally succeeds within 180 seconds, so kill it after 250
+    // seconds in case it is blocked forever due to unaddressed bugs.
+    return seastar::with_timeout(seastar::lowres_clock::now() + 220s, do_test(app))
```

Comment:
Make check is failed due to timeout.
```
...
INFO  2022-01-28 13:57:43,008 [shard 0] test - ms_handle_connect - Connection:[client.?(client) 127.0.0.1:0/69@58705 >> osd.55 v2:127.0.0.1:16822/55]
INFO  2022-01-28 13:57:43,414 [shard 0] test -  ms_dispatch conn= [client.?(client) 127.0.0.1:0/69@58705 >> osd.55 v2:127.0.0.1:16822/55] reply= i = 77
ERROR 2022-01-28 13:57:43,485 [shard 0] test - test_messenger timeout after 200s, abort! Consider to extend the period if the test is still running.
```
There is a global timeout for the make check job, so probably we can drop this?

Reply:
```
ERROR 2022-01-29 10:16:04,085 [shard 0] test - test_messenger timeout after 200s, abort! Consider to extend the period if the test is still running.
unittest-seastar-messenger-thrash: ../src/seastar/include/seastar/core/sharded.hh:563:
seastar::sharded<crimson::common::ConfigProxy>::~sharded() [T = crimson::common::ConfigProxy]: Assertion `_instances.empty()' failed.
```

Make check failure will be addressed by #44832, seems like `test_messenger_thrash` is also failing because `test_messenger` didn't end properly.

Reply:
> seems like test_messenger_thrash is also failing because test_messenger didn't end properly.

No, this log doesn't come from `test_messenger`, it's actually from `test_messenger_thrash` here. I suggested to drop this timeout from `test_messenger_thrash`.

Reply:
Make checks are ok now as timeout was increased, dropping it may cause and endless loop in `wait_for_done()`. Shouldn't we avoid it? (even by a relatively high timeout)

Reply:
There is a global timeout for the make check job (https://github.com/ceph/ceph-build/blob/master/ceph-pull-requests/build/build#L14).

For `test_messenger`, the manual timeout was added mainly due to
* the protocol interceptor that will add blocks intentionaly during the tests, and the timeout can terminate the incorrect block in time and print out the logs about the unexpected handshake steps.
* There was a historical issue that the `test_messenger` cannot be killed properly (should be fixed by https://github.com/ceph/ceph-build/pull/1764 already).